SignalMessaging/profiles/OWSUserProfile.m

```diff
@@ -0,0 +1,507 @@
+        if ([recipientId isEqualToString:kLocalProfileUniqueId]) {
+            [userProfile updateWithProfileKey:[OWSAES256Key generateRandomKey]
+                                 dbConnection:dbConnection
+                                   completion:nil];
+        }
+    }
+
+    OWSAssert(userProfile);
+
+    return userProfile;
+}
+
++ (BOOL)localUserProfileExists:(YapDatabaseConnection *)dbConnection
+{
+    __block BOOL result = NO;
+    [dbConnection readWithBlock:^(YapDatabaseReadTransaction *transaction) {
+        result = [OWSUserProfile fetchObjectWithUniqueID:kLocalProfileUniqueId transaction:transaction] != nil;
+    }];
+
+    return result;
+}
+
+- (instancetype)initWithRecipientId:(NSString *)recipientId
+{
+    self = [super initWithUniqueId:recipientId];
+
+    if (!self) {
+        return self;
+    }
+
+    OWSAssert(recipientId.length > 0);
+    _recipientId = recipientId;
+
+    return self;
+}
+
+#pragma mark - Update With... Methods
+
+- (void)finalizeWithCompletion:(nullable OWSUserProfileCompletion)externalCompletion
+{
+    DDLogVerbose(@"%@, after: %@", self.logTag, self.debugDescription);
+
+    if (externalCompletion) {
+        dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), externalCompletion);
+    }
+
+    BOOL isLocalUserProfile = [self.recipientId isEqualToString:kLocalProfileUniqueId];
+
+    dispatch_async(dispatch_get_main_queue(), ^{
+        if (isLocalUserProfile) {
+            // We populate an initial (empty) profile on launch of a new install, but until
+            // we have a registered account, syncing will fail (and there could not be any
+            // linked device to sync to at this point anyway).
+            if ([TSAccountManager isRegistered]) {
+                [CurrentAppContext() doMultiDeviceUpdateWithProfileKey:self.profileKey];
+            }
+
+            [[NSNotificationCenter defaultCenter] postNotificationNameAsync:kNSNotificationName_LocalProfileDidChange
+                                                                     object:nil
+                                                                   userInfo:nil];
+        } else {
+            [[NSNotificationCenter defaultCenter]
+                postNotificationNameAsync:kNSNotificationName_OtherUsersProfileWillChange
+                                   object:nil
+                                 userInfo:@{
+                                     kNSNotificationKey_ProfileRecipientId : self.recipientId,
+                                 }];
+            [[NSNotificationCenter defaultCenter]
+                postNotificationNameAsync:kNSNotificationName_OtherUsersProfileDidChange
+                                   object:nil
+                                 userInfo:@{
+                                     kNSNotificationKey_ProfileRecipientId : self.recipientId,
+                                 }];
+        }
+    });
+}
+
+- (void)updateWithProfileName:(nullable NSString *)profileName
```

Comment:
The `update with...` methods are incredibly repetitious but I don't see a good way to DRY them up.

We could `[MTLModel dictionaryValue]` to generate "before" and "after" snapshots and compare those, but we'd have to assume that we'll never run afoul of the class of "stale YDB model" bugs which the "update with..." pattern was intended to address.  Since we only use a single dedicated connection for `OWSUserProfile`, we should be safe, but it feels a little dangerous.

Reply:
Okay, I talked myself into taking the "model diffing" approach.